src/wgetopt.rs

```diff
@@ -369,251 +366,187 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
-                if nameend == p.name.len() {
-                    // The current option is exact match of this long option
-                    pfound = Some(*p);
-                    *indfound = option_index;
-                    *exact = true;
+            if potential_match
+                .name
+                .starts_with(&self.remaining_text[..name_end])
+            {
+                if name_end == potential_match.name.len() {
+                    // The option matches the text exactly, so we're finished.
+                    opt = Some(*potential_match);
+                    index = i;
+                    exact = true;
                     break;
-                } else if pfound.is_none() {
-                    // current option is first prefix match but not exact match
-                    pfound = Some(*p);
-                    *indfound = option_index;
+                } else if opt.is_none() {
+                    // The option begins with the matching text, but is not
+                    // exactly equal.
+                    opt = Some(*potential_match);
+                    index = i;
                 } else {
-                    // current option is second or later prefix match but not exact match
-                    *ambig = true;
+                    // There are multiple options that match the text non-exactly.
+                    ambiguous = true;
                 }
             }
         }
-        return pfound;
-    }
-
-    /// Check for a matching long opt.
-    fn _handle_long_opt(
-        &mut self,
-        longind: &mut usize,
-        long_only: bool,
-        retval: &mut char,
-    ) -> bool {
-        let mut exact = false;
-        let mut ambig = false;
-        let mut indfound: usize = 0;
-
-        let mut nameend = 0;
-        while self.nextchar.char_at(nameend) != '\0' && self.nextchar.char_at(nameend) != '=' {
-            nameend += 1;
-        }
-
-        let pfound = self._find_matching_long_opt(nameend, &mut exact, &mut ambig, &mut indfound);
 
-        if ambig && !exact {
-            self.nextchar = empty_wstr();
-            self.woptind += 1;
-            *retval = '?';
-            return true;
-        }
+        opt.map_or(LongOptMatch::NoMatch, |opt| {
+            if exact {
+                LongOptMatch::Exact(opt, index)
+            } else if ambiguous {
+                LongOptMatch::Ambiguous
+            } else {
+                LongOptMatch::NonExact(opt, index)
+            }
+        })
+    }
 
-        if let Some(pfound) = pfound {
-            self._update_long_opt(&pfound, nameend, longind, indfound, retval);
-            return true;
+    /// Check for a matching long-named option.
+    fn handle_long_opt(&mut self, longopt_index: &mut usize) -> Option<char> {
+        let name_end = self
+            .remaining_text
+            .chars()
+            .take_while(|c| !matches!(c, '\0' | '='))
+            .count();
```

Comment:
For reference the previous code was:

```
        let mut nameend = 0;
        while self.nextchar.char_at(nameend) != '\0' && self.nextchar.char_at(nameend) != '=' {
            nameend += 1;
        }
```

Note that `char_at` returns '\0' for one-past-the-end; we don't expect it to actually be in the option's name. So this should just find the position of any `=` sign; or `nameend.len()` if none.